Render test setup

ScottTodd opened the pull request.

benvanik commented on xrtl/testing/render_test.h:
```diff
@@ -0,0 +1,49 @@
+  // Returns false if the test image and golden image are different.
+  // The test image must have three components (rgb, 8 bits each) per pixel.
+  bool CompareImages(const std::string& golden_image_path,
+                     uint8_t* test_image_data, int test_image_width,
```
const void*

benvanik commented on xrtl/testing/render_test.h:
```diff
@@ -0,0 +1,49 @@
+//    auto pixel_data = ...;
+//    EXPECT_TRUE(CompareImages(path_to_golden, pixel_data, width, height));
+//  }
+class RenderTest : public ::testing::Test {
```
It'd be useful if RenderTest actually dealt with rendering and pixel comparison was orthogonal. That way things that have nothing to do with rendering (like image decoding/encoding, texture compression, etc) can use the pixel comparison. For that, testing/pixel_diff.h or testing/image_comparison.h or something could contain these comparison functions and be included from any test without forcing that test to subclass RenderTest.

benvanik commented on xrtl/testing/render_test.h:
```diff
@@ -0,0 +1,49 @@
+                     const std::string& test_image_path);
+
+  // Returns false if the test image and golden image are different.
+  // The test image must have three components (rgb, 8 bits each) per pixel.
```
This restriction is painful, as for example any framebuffer we read back will be 4 component rgba. We don't want to have to do format conversion in tests before calling this. Instead, having the function switch between a comparison of 1 component images against an alpha-only PNG, 3 component with RGB PNG, and 4 component with PNG with alpha would make it easier to use in a variety of scenarios.

benvanik commented on xrtl/testing/render_test.cc:
```diff
@@ -0,0 +1,98 @@
+  if (test_image_width != golden_image_width) {
+    LOG(ERROR) << "Image widths don't match. Expected " << golden_image_width
+               << "px but test was " << test_image_width << "px";
+    stbi_image_free(golden_image_data);
```
you could wrap the data in a unique_ptr to avoid the need for the manual memory tracking. for example:

```
std::unique_ptr<uint8_t, void(*)(uint8_t*)> golden_image_data{stbi_load(...), stbi_image_free};
```

Could make this a helper function so you can just auto golden_image_data = ...;

benvanik commented on xrtl/gfx/pixel_diff_test.cc:
```diff
@@ -0,0 +1,42 @@
+// Checks that pixel diffing render tests work.
+class PixelDiffTest : public testing::RenderTest {};
+
+TEST_F(PixelDiffTest, SameImage) {
```
move these side-by-side with the CompareImages code (so if you had testing/pixel_diff.h/cc, this would be pixel_diff_test.cc)

ScottTodd commented:
Addressed comments by splitting RenderTest into ImageComparison and ImageLoader under testing/. Can later add more infrastructure to generate and display diffs, approve changes, etc.

Still probably doesn't work on Windows without a runfiles fix or workaround.

ScottTodd commented:
New tests pass on Windows now
